BUGFIX: Failed validation prevents resource from being persisted

albe opened the pull request:
This is achieved by deferring the importing and persisting of resources until after validation.

Fixes #769

albe committed:
BUGFIX: Failed validation prevents resource from being persisted

This is achieved by deferring the importing and persisting of resources until after validation.

albe committed:
TASK: Lazy evaluation of deferred resource sha1, md5 and filesize

albe committed:
TASK: Some docblock type fixes

albe committed:
BUGFIX: FileSystemStorage::getObjects correctly yields single StorageObjects

albe committed:
TASK: Slight restructuring

albe committed:
Apply fixes from StyleCI

albe commented:
When reviewing, please take a look at the commit https://github.com/neos/flow-development-collection/pull/1931/commits/5120bd76c14fd6bdaf4559ee51f7c50ff147555e first - it only includes the functional changes required for this. All others are only distracting, but useful docblock type changes.

albe commented on Neos.Flow/Classes/Mvc/Controller/AbstractController.php:
```diff
@@ -12,6 +12,8 @@
  */
 
 use GuzzleHttp\Psr7\Uri;
+use Neos\Flow\ResourceManagement\PersistentResource;
```
This sucks, but is required unless someone else has a good idea how to trigger resource importing **after** validation but before controller invokation.

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -285,6 +322,14 @@ public function setFileSize($fileSize)
      */
     public function getFileSize()
     {
+        if (!$this->fileSize && $this->source !== null) {
```
This (and below lazy evaluations) is maybe not needed, but I'm unsure if the resource properties could be accessed before it being "imported" via the resource manager

albe committed:
TASK: Adjust tests and make import exception a validation error again

albe committed:
Apply fixes from StyleCI

albe commented on Neos.Flow/Classes/Mvc/Controller/AbstractController.php:
```diff
@@ -374,10 +383,23 @@ protected function throwStatus($statusCode, $statusMessage = null, $content = nu
+                    try {
+                        $this->resourceManager->persistResource($value);
+                    } catch (\Exception $exception) {
+                        $this->logger->warning('Could not import an uploaded file', ['exception' => $exception] + LogEnvironment::fromMethodName(__METHOD__));
```
Needed to keep the behaviour of import exceptions ending up as validation errors

kdambekalns commented:
What about `importResource()`, it is obviously left unchanged in this PR. How do those changes affect that? Do they? Why and why not? I think there should be some explanation on that, ideally close to the code involved…

albe commented:
> What about importResource()?

That's a fair question. It (and sibling methods) are indeed unchanged, because an actual *import* of a Resource should make it persistent (in storage) immediately. Then again, those methods take arbitrary forms of data and convert that into a `PersistentResource`, while the new `persistResource` method takes a (deferred) Resource and only makes it persistent. Also, this makes sure b/c is guaranteed.

Now in regard to where to place this information closer to the code - I'll give it a thought and add something. Thanks for the feedback!

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -125,6 +133,14 @@ class PersistentResource implements ResourceMetaDataInterface, CacheAwareInterfa
     protected $temporaryLocalCopyPathAndFilename;
 
+    /**
+     * @param string|resource|null $source
```
```suggestion
     * The source is required for deferred persistence (after validation) in the framework and should otherwise not be set.
     *
     * @param string|resource|null $source
```

albe commented on Neos.Flow/Classes/ResourceManagement/ResourceManager.php:
```diff
@@ -260,6 +260,40 @@ public function importUploadedResource(array $uploadInfo, $collectionName = self
     }
 
+    /**
+     * Import a deferred resource into it's destination collection and persist it.
```
```suggestion
     * Import a deferred resource into it's destination collection and persist it.
     * This will do nothing, if the resource is not deferred (has no source attached).
```

albe commented on Neos.Flow/Classes/ResourceManagement/ResourceManager.php:
```diff
@@ -260,6 +260,40 @@ public function importUploadedResource(array $uploadInfo, $collectionName = self
+        }
+
+        $this->resourceRepository->add($resource);
+        $this->logger->debug(sprintf('Successfully imported the uploaded file "%s" into the resource collection "%s" (storage: "%s", a %s. SHA1: %s)', $resource->getFilename(), $collectionName, $this->collections[$collectionName]->getStorage()->getName(), get_class($this->collections[$collectionName]->getStorage()), $resource->getSha1()));
```
```suggestion
        $this->logger->debug(sprintf('Successfully imported the uploaded file "%s" into the resource collection "%s" (storage: "%s", a %s. SHA1: %s)', $resource->getFilename(), $collectionName, $collection->getStorage()->getName(), get_class($collection->getStorage()), $resource->getSha1()));
```

albe committed:
TASK: Improve doc blocks

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -125,6 +133,16 @@ class PersistentResource implements ResourceMetaDataInterface, CacheAwareInterfa
+    /**
+     * The source is required for deferred persistence (after validation) in the framework and should otherwise not be set.
+     *
+     * @param string|resource|null $source
```
```suggestion
     * @param string|array|resource|null $source
```

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -36,6 +36,14 @@ class PersistentResource implements ResourceMetaDataInterface, CacheAwareInterfa
+    /**
+     * The source for the resource when being newly created for deferred persistence
+     *
+     * @var string|resource|null
```
```suggestion
     * @var string|array|resource|null
```

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -143,14 +161,35 @@ public function initializeObject($initializationCause)
     }
 
+    /**
+     * @return resource|string|null
```
```suggestion
     * @return resource|array|string|null
```

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -143,14 +161,35 @@ public function initializeObject($initializationCause)
+     * Detach the source from this resource and return it.
+     * Used for deferred persistence.
+     *
+     * @return resource|string|null
```
```suggestion
     * @return resource|array|string|null
```

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -285,6 +324,14 @@ public function setFileSize($fileSize)
+        if (!$this->fileSize && $this->source !== null) {
+            if (is_string($this->source)) {
+                $this->fileSize = filesize($this->source) ?: 0;
+            } elseif (is_resource($this->source)) {
```
```suggestion
            } elseif (is_array($this->source)) {
                $this->fileSize = $this->source['size'] ?? 0;
            } elseif (is_resource($this->source)) {
```

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -296,6 +343,16 @@ public function getFileSize()
+        if (!$this->sha1 && $this->source !== null) {
+            if (is_string($this->source)) {
+                $this->sha1 = sha1_file($this->source);
+            } elseif (is_resource($this->source)) {
```
```suggestion
            } elseif (is_array($this->source) && isset($this->source['tmp_name'])) {
                $this->sha1 = sha1_file($this->source['tmp_name']);
            } elseif (is_resource($this->source)) {
```

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -323,6 +380,16 @@ public function setSha1($sha1)
+        if (!$this->md5 && $this->source !== null) {
+            if (is_string($this->source)) {
+                $this->md5 = md5_file($this->source);
+            } elseif (is_resource($this->source)) {
```
```suggestion
            } elseif (is_array($this->source) && isset($this->source['tmp_name'])) {
                $this->md5 = md5_file($this->source['tmp_name']);
            } elseif (is_resource($this->source)) {
```

albe commented on Neos.Flow/Classes/ResourceManagement/ResourceManager.php:
```diff
@@ -260,6 +260,41 @@ public function importUploadedResource(array $uploadInfo, $collectionName = self
+        $collection = $this->collections[$collectionName];
+        try {
+            // Optimally the collection would just accept a PersistentResource and update it, but this would mean an API change
+            $importedResource = $collection->importResource($resource->detachSource());
```
```suggestion
            $source = $resource->detachSource();
            if (is_array($source)) {
                $importedResource = $collection->importUploadedResource($source, $collectionName);
            } else {
                $importedResource = $collection->importResource($source, $collectionName);
            }
```

albe committed:
TASK: Correctly handle PHP file uploads

This is probably obsolete since PSR-7 FileUpload anyway.

albe commented:
After thinking around I can't seem to find a good place or wording to describe the non-effect on the import* methods, other then only describing the "deferred" concept better in the `PersistentResource` constructor: it is only intended for framework purposes and should not be manually used (unless you know what you do). So I'm happy for any ideas.

Noticed I missed handling arrays correctly and fixed that, but I think no upload ends up as an array in the converter any more since the PSR-7 change. Have to check that - maybe we can get rid of that for next major.

kitsunet commented on Neos.Flow/Classes/ResourceManagement/ResourceTypeConverter.php:
```diff
@@ -292,26 +311,28 @@ protected function handleUploadedFile(UploadedFileInterface $source, PropertyMap
 
         try {
-            $resource = $this->resourceManager->importResource($source->getStream()->detach(), $this->getCollectionName($source, $configuration));
+            $resource = new PersistentResource($source->getStream()->detach());
```
That's a (kind of big) behavior change isn't it? If I use this converter I could so far expect it to create a persisted resource.

albe replied:
Well, it is an `"eventually persisted resource"` so to say. I'd be happy to find another solution to allow validation to prevent persisting (and importing) a resource, but currently the property mapping is done *before* validation. And this is not a feature, but actually a bug. A Resource that doesn't pass validation should not end up in the database, let alone on persistent storage. But it's also not easy to just remove the resource from persistence again after validation, as you'd need to know if the resource was newly added or just updated (plus the unnecessary work done). So the converter really should not call `repository->add()` (the `PersistentObjectConverter` doesn't either), nor move the resource to a storage backend.

The only valid other option I see would be to split validation in two separate steps, then only have the first step allow rejecting resources (in their "source" form). But that would be one hell of a change in regards to code required, complexity, breakyness and required documentation to make people understand what the two validations are.

Anyway: Thanks for challenging the solution!

albe replied:
PS: This whole converter is the reason I did not introduce a `EventuallyPersistedResource extends PersistentResource` class - because the converter would need to change, which would be super-breaking. With my solution this is more subtle and a `PersistentResource` is differentiated by the `source` property. One with a source has definitely not yet been imported! It of course won't work if you manually create a `new PersistentResource($someSource)` and add that to a repo. But that also didn't work before, you needed to use the `ResourceManager` and if you do that, this solution right now also works.
We could maybe also add a safety-check into the `ResourceRepository` that allows only adding resources that have no source attached, but then again, that would only make it better for those cases where you specify a `$source` - which in turn would be fine in scope of this change.

kitsunet commented on Neos.Flow/Classes/ResourceManagement/ResourceTypeConverter.php:
```diff
@@ -250,38 +260,47 @@ protected function handleHashAndData(array $source, PropertyMappingConfiguration
-                if (is_array($source) && isset($source['filename'])) {
-                    $resource->setFilename($source['filename']);
-                }
+                $resource = new PersistentResource($configuration->getConfigurationValue(ResourceTypeConverter::class, self::CONFIGURATION_RESOURCE_LOAD_PATH) . '/' . $hash);
```
Same as other comment, regarding behavior change.

kitsunet commented on Neos.Flow/Classes/ResourceManagement/ResourceTypeConverter.php:
```diff
@@ -207,12 +210,19 @@ protected function handleFileUploads(array $source, PropertyMappingConfiguration
 
         try {
-            $resource = $this->resourceManager->importUploadedResource($source, $this->getCollectionName($source, $configuration));
+            $resource = new PersistentResource($source);
```
and third time ;)

kitsunet commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -125,6 +133,16 @@ class PersistentResource implements ResourceMetaDataInterface, CacheAwareInterfa
+     *
+     * @param string|array|resource|null $source
+     */
+    public function __construct($source = null)
```
ooof, especially with the multi type I am not a huge fan of this. ESPECIALLY as this thing is called persisted resource and it is not persisted at all if this is used with the temporary files. In fact someone could come up with the silly idea of actually creating a PersistedResource this way and just handing it to the repository (which obviously fails miserably). I don't like this change and would rather introduce something else for this use case...

albe replied:
I see your point here. What about a static constructor method that creates such a deferred persistent resource? Generally, you should not manually create such a deferred resource (you probably should never create a `PersistentResource` manually at all).

kitsunet reviewed (changes requested):
I am flagging this as request changes, I am really not happy with this change as is and I hope we find a better solutoin (which might become a feature then...)

albe commented:
ping @kitsunet - did you have time to give it another thought as we discussed?

albe committed:
TASK: Improve deferred persistent resource implementation

- use a static factory instead of constructor with optional argument
- prevent ResourceRepository from accepting deferred resources
- replace `getSource()` with `isDeferred()` getter, since you should never get the source without detaching

albe commented on Neos.Flow/Classes/ResourceManagement/ResourceRepository.php:
```diff
@@ -74,6 +74,9 @@ public function __construct()
      */
     public function add($object)
     {
+        if ($object instanceof PersistentResource && $object->isDeferred()) {
```
@kitsunet see my other comment - I think this is a good measure to prevent the primary line of accidential misuse

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -125,6 +133,19 @@ class PersistentResource implements ResourceMetaDataInterface, CacheAwareInterfa
+     *
+     * @param string|array|resource $source
+     */
+    public static function createDeferred($source)
```
This also feels better than the optional constructor argument. Should have done that from the start!

albe commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -143,14 +164,35 @@ public function initializeObject($initializationCause)
+    /**
+     * @return bool True if this resource is deferred and has a source attached that needs to be imported first.
+     */
+    public function isDeferred(): bool
```
Also better than the getter that was only used to check if source is null

albe commented:
As said in our unicorn weekly I even think this could be considered a security issue, in which case we would need to target 4.3 - I'd take care if we agree on this.

@kitsunet @kdambekalns @robertlemke @bwaidelich @sorenmalling @daniellienert (sorry for the roundhouse mention)

albe commented:
ping y'all

sorenmalling commented on Neos.Flow/Classes/ResourceManagement/PersistentResource.php:
```diff
@@ -323,6 +387,18 @@ public function setSha1($sha1)
      */
     public function getMd5()
```
We removed the md5 part of PersistenceResource #2112

albe replied:
So since this targets 5.3 (as it is a BUGFIX), we'll need to have an immediate follow-up for master (or carefully resolve the removal in the up-merge conflict then). Thanks for the heads-up! :)

sorenmalling commented:
It's a solution to the original issue, and as you write

> PS: And yes, this is not an optimal solution. Let's keep improving this in master further to avoid the "deffered persisted resource" paradox, but this will inevitably be a breaking change then. 

Did our possibilities change with the introduction of PSR-15 - can we handle a GC of uploaded resources, that is up for deletion due to validation failure once the request "is done"?

albe commented:
> can we handle a GC of uploaded resources, that is up for deletion due to validation failure once the request "is done"?

The core issue is really that the resource is added for persistence before validation in the first place. It's non-trivial to "un-add" cleanly. And the fact that validation is only run directly before invoking the action method inside the controller, so there is not much room for "in between" layer that does this (after the action is too late, because the action might call `persistAll()`).

albe commented on Neos.Flow/Classes/ResourceManagement/Storage/FileSystemStorage.php:
```diff
@@ -131,12 +131,12 @@ public function getStreamByResourcePath($relativePath)
     {
         foreach ($this->resourceManager->getCollectionsByStorage($this) as $collection) {
-            yield $this->getObjectsByCollection($collection, $callback);
+            yield from $this->getObjectsByCollection($collection, $callback);
```
Note: This is actually another bug looming in the code right now, no?

albe replied:
See #2167

albe commented on Neos.Flow/Classes/ResourceManagement/ResourceTypeConverter.php:
```diff
@@ -237,7 +247,7 @@ protected function handleHashAndData(array $source, PropertyMappingConfiguration
             }
 
-            if ($configuration->getConfigurationValue(ResourceTypeConverter::class, self::CONFIGURATION_IDENTITY_CREATION_ALLOWED) !== true) {
+            if ($configuration === null || $configuration->getConfigurationValue(ResourceTypeConverter::class, self::CONFIGURATION_IDENTITY_CREATION_ALLOWED) !== true) {
```
Note: Maybe a missing configuration should not allow identity creation for the resource?

kdambekalns replied:
Right now that is the behaviour, right? So changing it would make this a breaking change?

albe replied:
Well, right now if no configuration is given it just fails miserably. It's not a case that will happen through the property mapper (that one requires a configuration), only if typeconverters are invoked manually.

kdambekalns replied:
Ok, in that case I am all for "deny unless explicitly configured".

albe replied:
```suggestion
            if ($configuration !== null && $configuration->getConfigurationValue(ResourceTypeConverter::class, self::CONFIGURATION_IDENTITY_CREATION_ALLOWED) !== true) {
```

albe replied:
Fixed that separately with #2349

albe commented:
Rebased to keep things in sync

albe commented:
ping @kdambekalns @robertlemke @daniellienert @bwaidelich still would need someone of you to have a close look and approve or reject this. Maybe @kitsunet could remove the veto unless the concerns are still not alleviated

albe commented:
Okay everyone - I had an idea for an alternative implementation, that will not introduce a new concept of a `PersistentResource` and be much less breaking. It is partly inspired by @sorenmalling's question here and @bwaidelich's `onStep()` callback implementation we used for PSR-15 middleware chain to hook into the steps to update our RequestHandler's `$request` instance. So props to them both.

Please take a look there - if we can agree on that, this can be closed for good.

albe committed:
Merge branch '5.3' into persistent-resources-defer

albe committed:
Merge branch '5.3' into persistent-resources-defer

# Conflicts:
#	Neos.Flow/Classes/ResourceManagement/ResourceTypeConverter.php

albe commented on Neos.Flow/Classes/ResourceManagement/Storage/FileSystemStorage.php:
```diff
@@ -158,7 +158,7 @@ public function getObjectsByCollection(CollectionInterface $collection, callable
             $object->setMd5($resource->getMd5());
             $object->setFileSize($resource->getFileSize());
-            $object->setStream(function () use ($resource) {
+            $object->setStream(static function () use ($resource) {
```
```suggestion
            $object->setStream(function () use ($resource) {
```

albe committed:
TASK: Remove wrong usage of static anonymous function

fcool commented:
Just some general thoughts here: We had this in several Pen-Tests reported to be an issue (importing a file into local storage, even if not validated).
Our approach has been to write a validating ResourceConverter, which can be extended in some way, to make the testers happy. 

In general I really dislike the behavior that the database gets modified by a TypeConverter - persisting changes should be a task the controller has full control of - so I would love a solution, which do not persist the resource already. But I understand why this seems "desired", when it comes to resources. But it leads to all sort of "unexpected" behavior. For example just receiving a "file object" you will only work with ONCE in your controllers action will be saved in storage and database and rot there unrelated. Something we should probably at least address in documentation.

albe commented:
Yeah, the best solution would be if a resource is not persisted at all (neither storage, nor DB) until the controller decides it wants to. The storage persistence could be solved by a onFlush listener and the controller would need to do `$this->persistenceManager->persist($uploadedResource)` (or persist the parent entity). That would be a breaking change though, but maybe something for Flow 8?

Anyway, regarding this BUGFIX - as mentioned in the other PR I'm leaning a bit more towards this solution right now, but with a few adjustments. And I agree to documenting the behaviour - maybe with a suggestion on how to deal with "temporary" file uploads properly. Hopefully after the 7.2 release I can find some time to continue the work on this.

kdambekalns reviewed (changes requested):
I favour the other PR as a fix!
